pipeline/execout/cachev1/engine.go

```diff
@@ -0,0 +1,143 @@
+package cachev1
+
+import (
+	"context"
+	"fmt"
+	"github.com/streamingfast/bstream"
+	"github.com/streamingfast/dstore"
+	pbsubstreams "github.com/streamingfast/substreams/pb/sf/substreams/v1"
+	"github.com/streamingfast/substreams/pipeline/execout"
+	"go.uber.org/zap"
+	"regexp"
+)
+
+var cacheFilenameRegex *regexp.Regexp
+
+func init() {
+	cacheFilenameRegex = regexp.MustCompile(`([\d]+)-([\d]+)\.output`)
+}
+
+type Engine struct {
+	ctx               context.Context
+	caches            map[string]*OutputCacheState
+	SaveBlockInterval uint64
+	baseCacheStore    dstore.Store
+	logger            *zap.Logger
+}
+
+type OutputCacheState struct {
+	c           *OutputCache
+	initialized bool
+}
+
+func NewEngine(ctx context.Context, saveBlockInterval uint64, baseCacheStore dstore.Store, logger *zap.Logger) (execout.CacheEngine, error) {
+	e := &Engine{
+		ctx:               ctx,
+		caches:            make(map[string]*OutputCacheState),
+		SaveBlockInterval: saveBlockInterval,
+		baseCacheStore:    baseCacheStore,
+		logger:            logger,
+	}
+	return e, nil
+}
+
+func (e *Engine) Prime(modules map[string]*pbsubstreams.Module) error {
+	for hash, module := range modules {
+		if err := e.registerCache(module, hash); err != nil {
+			return fmt.Errorf("failed to register chache for module %q: %w", module.Name, err)
+		}
+	}
+	return nil
+}
+
+func (e *Engine) FlushAndUpdate(ctx context.Context, blockRef bstream.BlockRef) error {
+	e.logger.Info("flushing output caches and updating")
+	for _, moduleCache := range e.caches {
+		if moduleCache.c.IsOutOfRange(blockRef) {
+			if err := e.flushCache(ctx, moduleCache.c); err != nil {
+				return fmt.Errorf("failed to flush cache: %w", err)
+			}
+
+			if _, err := moduleCache.c.LoadAtBlock(ctx, moduleCache.c.CurrentBlockRange.ExclusiveEndBlock); err != nil {
+				return fmt.Errorf("loading blocks %d for module kv %s: %w", moduleCache.c.CurrentBlockRange.ExclusiveEndBlock, moduleCache.c.ModuleName, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (e *Engine) Flush(ctx context.Context) error {
+	e.logger.Info("flushing output caches")
+	for _, moduleCache := range e.caches {
+		if err := e.flushCache(ctx, moduleCache.c); err != nil {
+			return fmt.Errorf("failed to flush cache: %w", err)
+		}
+	}
+	return nil
+}
+
+func (e *Engine) UndoBlock(moduleName string, blockID string) {
+	if outputCache, found := e.caches[moduleName]; found {
+		outputCache.c.Delete(blockID)
+	}
+}
+
+func (e *Engine) flushCache(ctx context.Context, cache *OutputCache) error {
+	filename := cache.currentFilename()
+	e.logger.Debug("saving cache",
+		zap.String("module_name", cache.ModuleName),
+		zap.Uint64("start_block", cache.CurrentBlockRange.StartBlock),
+		zap.Uint64("end_block", cache.CurrentBlockRange.ExclusiveEndBlock),
+	)
+
+	if err := cache.save(ctx, filename); err != nil {
+		return fmt.Errorf("save: saving outpust or module kv %s: %w", cache.ModuleName, err)
+	}
+	return nil
+}
```

Comment:
There's an important issue here we discussed but still isn't fixed. Perhaps it's documented somewhere else, but I'm dropping my notes here in case it isn't:

This function writes the output cache NO MATTER what state of the chain we're at. We should not save any files unless we're certain we've passed finality on all the blocked we added through `NewBlock()`.

We either need to be informed of finality through the steps handler of our ForkHandler (a custom one with all the steps).

Maybe this means therefore, that we should not have a `NewBlock()` but rather a `NewFinalBlock()` and make sure we never call this when we only have a `STEP_NEW`.  We don't want to fiddle with the cache when we're neaer real-time anyway, do we?!